README.md
RNSchrittmotor
--------------
http://github.com/slaxor/rnschrittmotor

DESCRIPTION:
------------
This is a convenience lib for the Stepper-Controller board from
[robotikhardware shop](http://www.shop.robotikhardware.de/shop/catalog/product_info.php?products_id=173 "Robotikhardware Shop RN-Schrittmotor"). It follows this
[API-Doc](http://www.robotikhardware.de/download/rnschrittmotor.pdf "RN-Schrittmotor API-Doc")

SYNOPSIS:
---------
Use it like this:

    s = RNSchrittmotor.new('/dev/ttyUSB0', 9600)
    s.read_version
    s.set_crc_mode(RNSchrittmotor::CRCON)
    s.step!(RNSchrittmotor::BOTH, 400)
    # RTFS for more

REQUIREMENTS:
-------------
It currently depends on serialport

INSTALL:
--------
    gem install rnschrittmotor

LICENSE:
--------
© 2010 Sascha Teske <sascha.teske@microprojects.de>
General Public License version 3 or later (in this folder gpl_v3.txt)



CHANGELOG
0.2 (May 13, 2010)
------------------
* First simple release with complete API implementation



test/test_helper.rb
require 'rubygems'
require 'test/unit'
require 'mocha'
require 'stringio'
require File.join([File.expand_path(File.dirname(__FILE__)), '..', 'lib', 'rn_schrittmotor'])



test/rn_schrittmotor_test.rb
require File.dirname(__FILE__) + '/test_helper.rb'

class RNSchrittmotorTest < Test::Unit::TestCase

  def setup
    @stepper = RNSchrittmotor.new('/dev/ttyUSB0', 9600)
  end

  def test_crc8
    assert_equal "\x56",  RNSchrittmotor.crc8("\xff\x00\x00\x00\x00\x00")
    assert_equal "\x0e",  RNSchrittmotor.crc8("\xfe\x0f\x00\x00\x00\x00")
  end

  def test_do_it
    @stepper.expects(:write).with("!#\xff\x00\x00\x00\x00\x00\x56")
    @stepper.expects(:read)
    @stepper.do_it("\xff\x00\x00\x00\x00\x00")
  end

  #eeprom_bytes
  #0 I2C Slave ID
  #1 Motorstrom in mA Motor 1 Low-Byte
  #2 Motorstrom in mA Motor 1 High-Byte
  #3 Motorstrom in mA Motor 2 Low-Byte
  #4 Motorstrom in mA Motor 2 High-Byte
  #5 Haltestrom in mA Motor 1 Low-Byte
  #6 Haltestrom in mA Motor 1 High-Byte
  #7 Haltestrom in mA Motor 2 Low-Byte
  #8 Haltestrom in mA Motor 2 High-Byte
  #9 Anlaufstrom in mA Motor 1 Low-Byte
  #10 Anlaufstrom in mA Motor 1 High-Byte
  #11 Anlaufstrom in mA Motor 2 Low-Byte
  #12 Anlaufstrom in mA Motor 2 High-Byte
  #13 Schrittmotormodus 0=Vollschritt 1=Halbschritt
  #14 Schnittstellenmodus 1=Intelligent 2=Takt
  #15 CRC Mode 1=aktiv 0=aus

  def test_export_settings
    @stepper.expects(:do_it).with("\xfe\x10\x00\x00\x00\x00").returns("\x42\x00\x04\x00\x04\x00\x05\x00\x05\xfa\x00\xfa\x00\x01\x02\x00")
    assert_equal %Q(i2c_slave_id = 66
stepper_current = [1024, 1024]
start_current = [1280, 1280]
holding_current = [250, 250]
stepping_mode = 1
controller_mode = 2
crc_mode = 0
), @stepper.export_settings
  end

  def test_number_to_bytes
    assert_equal "\x04\x0c", RNSchrittmotor.number_to_bytes(3076)
  end

  def test_bytes_to_number
    assert_equal 32515, RNSchrittmotor.bytes_to_number("\x03\x7f")
  end

  def test_read_counter
    @stepper.expects(:do_it).with("f\001\000\000\000\000").returns("\xff\xff")
    assert_equal "\xff\xff", @stepper.read_counter(RNSchrittmotor::FIRST)
  end

  def test_read_eeprom
    @stepper.expects(:do_it).with("\xfe\x20\000\000\000\000").returns("\xff" * 32)
    assert_equal "\xff" * 32, @stepper.read_eeprom(32)
  end

  def test_read_end_switches
    @stepper.expects(:do_it).with("h\000\000\000\000\000").returns("\x0f")
    assert_equal "\x0f", @stepper.read_end_switches
  end

  def test_read_i2c_response
    @stepper.expects(:do_it).with("g\000\000\000\000\000").returns("*")
    assert_equal "*", @stepper.read_i2c_response
  end

  def test_read_status
    @stepper.expects(:do_it).with("e\001\000\000\000\000").returns("\x00")
    assert_equal "\x00", @stepper.read_status(RNSchrittmotor::FIRST)
  end

  def test_read_version
    @stepper.expects(:do_it).with("\xff\000\000\000\000\000").returns("blah version string")
    assert_equal "blah version string", @stepper.read_version
  end

  def test_reset_counter
    @stepper.expects(:do_it).with("\016\001\000\000\000\000").returns("*")
    assert_equal "*", @stepper.reset_counter(RNSchrittmotor::FIRST)
  end

  def test_reset_eeprom
    @stepper.expects(:do_it).with("\313\313\000\000\000\000").returns("*")
    assert_equal "*", @stepper.reset_eeprom(203)
  end

  def test_run!
    @stepper.expects(:do_it).with("6\001\000\000\000\000").returns("*")
    assert_equal "*", @stepper.run!(RNSchrittmotor::FIRST)
  end

  def test_set_controller_mode
    @stepper.expects(:do_it).with("\310\000\000\000\000\000").returns("*")
    assert_equal "*", @stepper.set_controller_mode(RNSchrittmotor::SERIAL)
    @stepper.expects(:do_it).with("\310\000\000\000\000\000").returns("*")
    assert_equal "*", @stepper.set_controller_mode(RNSchrittmotor::I2C)
    @stepper.expects(:do_it).with("\310\002\000\000\000\000").returns("*")
    assert_equal "*", @stepper.set_controller_mode(RNSchrittmotor::PULSE)
  end

  def test_set_crc_mode
    @stepper.expects(:do_it).with("\311\001\000\000\000\000").returns("*")
    assert_equal "*", @stepper.set_crc_mode(RNSchrittmotor::CRCON)
    @stepper.expects(:do_it).with("\311\000\000\000\000\000").returns("*")
    assert_equal "*", @stepper.set_crc_mode(RNSchrittmotor::CRCOFF)
  end

  def test_set_direction
    @stepper.expects(:do_it).with("4\001\000\000\000\000").returns("*")
    assert_equal "*", @stepper.set_direction(RNSchrittmotor::FIRST, RNSchrittmotor::LEFT)
    @stepper.expects(:do_it).with("4\001\000\000\000\000").returns("*")
    assert_equal "*", @stepper.set_direction(RNSchrittmotor::FIRST, RNSchrittmotor::CCW)
    @stepper.expects(:do_it).with("4\001\001\000\000\000").returns("*")
    assert_equal "*", @stepper.set_direction(RNSchrittmotor::FIRST, RNSchrittmotor::RIGHT)
    @stepper.expects(:do_it).with("4\001\001\000\000\000").returns("*")
    assert_equal "*", @stepper.set_direction(RNSchrittmotor::FIRST, RNSchrittmotor::CW)
  end

  def test_set_i2c_slave_id
    @stepper.expects(:do_it).with("\312\026\000\000\000\000").returns("*")
    assert_equal "*", @stepper.set_i2c_slave_id(22)
  end

  def test_steps_per_sec
     assert_equal 0, RNSchrittmotor.steps_per_sec(1000)
     assert_equal 1, RNSchrittmotor.steps_per_sec(500)
     assert_equal 250, RNSchrittmotor.steps_per_sec(3)
  end

  def test_set_speed
    @stepper.expects(:do_it).with("5\001\377\377\000\000").returns("*")
    assert_equal "*", @stepper.set_speed(RNSchrittmotor::FIRST, 255, 255)
    @stepper.expects(:do_it).with("5\001\004\005\000\000").returns("*")
    assert_equal "*", @stepper.set_speed(RNSchrittmotor::FIRST, RNSchrittmotor.steps_per_sec(200), 5)
  end

  def test_set_start_current
    @stepper.expects(:do_it).with("\x0b\003\x08\x07\x00\x00").returns("*")
    assert_equal "*", @stepper.set_start_current(RNSchrittmotor::BOTH, 1800, RNSchrittmotor::FORGET)
  end

  def test_set_stepper_current
    @stepper.expects(:do_it).with("\x0a\x01\xe8\x03\x00\x00").returns("*")
    assert_equal "*", @stepper.set_stepper_current(RNSchrittmotor::FIRST, 1000, RNSchrittmotor::FORGET)
  end

  def test_set_holding_current
    @stepper.expects(:do_it).with("\x0c\x01\x64\000\000\000").returns("*")
    assert_equal "*", @stepper.set_holding_current(RNSchrittmotor::FIRST, 100, RNSchrittmotor::FORGET)
    @stepper.expects(:do_it).with("\x0c\002\x64\000\001\000").returns("*")
    assert_equal "*", @stepper.set_holding_current(RNSchrittmotor::SECOND, 100, RNSchrittmotor::PERSIST)
  end

  def test_set_stepping_mode
    @stepper.expects(:do_it).with("\r\001\000\000\000\000").returns("*")
    assert_equal "*", @stepper.set_stepping_mode(RNSchrittmotor::FULL, RNSchrittmotor::FORGET)
    @stepper.expects(:do_it).with("\r\000\001\000\000\000").returns("*")
    assert_equal "*", @stepper.set_stepping_mode(RNSchrittmotor::HALF, RNSchrittmotor::PERSIST)
  end

  def test_step!
    @stepper.expects(:do_it).with("7\001\270\v\000\000").returns("*")
    assert_equal "*", @stepper.step!(RNSchrittmotor::FIRST, 3000)
  end

  def test_switch_off!
    @stepper.expects(:do_it).with("3\001\000\000\000\000").returns("*")
    assert_equal "*", @stepper.switch_off!(RNSchrittmotor::FIRST)
  end

  def test_switch_on!
    @stepper.expects(:do_it).with("2\001\000\000\000\000").returns("*")
    assert_equal "*", @stepper.switch_on!(RNSchrittmotor::FIRST)
  end

  def test_emergency_off!
    @stepper.expects(:do_it).with("3\003\000\000\000\000").returns("*")
    assert_equal "*", @stepper.emergency_off!
  end
end



lib/rn_schrittmotor.rb
#!/usr/bin/env ruby
require 'serialport'
class RNSchrittmotor < SerialPort

  @@debug = false

  SERIAL = I2C = 0
  PULSE = 2

  FIRST = 1
  SECOND = 2
  BOTH = 3

  CCW = LEFT = 0
  CW = RIGHT = 1

  CRCOFF = 0
  CRCON = 1

  HALF = 0
  FULL = 1

  FORGET = 0
  PERSIST = 1

  RETURNCODES = { '*' => 'OK', '+' => 'Unknown Command', ',' => 'Wrong CRC', '-' => 'Wrong/Odd Slave ID'}

  COMMANDS = [
    [:set_stepper_current,   10, :motor, :mA, :persist],
    [:set_start_current,     11, :motor, :mA, :persist],
    [:set_holding_current,   12, :motor, :mA, :persist],
    [:set_stepping_mode,     13, :mode, :persist],
    [:reset_counter,         14, :motor],
    [:switch_on!,            50, :motor], # are you current settings correct for you stepper?
    [:switch_off!,           51, :motor], # you might wanna be careful if eg. your z-axis is high up it just falls down
    [:set_direction,         52, :motor, :direction],
    [:set_speed,             53, :motor, :speed, :accel],
    [:run!,                  54, :motor],
    [:step!,                 55, :motor, :steps],
    [:read_status,          101, :motor],
    [:read_counter,         102, :motor],
    [:read_i2c_response,    103],
    [:read_end_switches,    104],
    [:set_controller_mode,  200, :mode],
    [:set_crc_mode,         201, :onoff],
    [:set_i2c_slave_id,     202, :i2c_id],
    [:reset_eeprom,         203, :confirm], # totally braindead
    [:read_eeprom,          254, :size],
    [:read_version,         255]
  ]

  COMMANDS.each do |command_name, command_byte, *params|
    class_eval %Q(
      def #{command_name}(#{params.join(',')})
        cmd = #{command_byte}.chr
        #{params.map {|p| [:mA, :steps].include?(p) ? "cmd << self.class.number_to_bytes(" + p.to_s + ")\n" : "cmd << " + p.to_s + ".chr\n"}}
        do_it(cmd.ljust(6,0.chr))
      end
    )
  end

  def emergency_off!
   switch_off!(BOTH)
  end

  def export_settings
    ee = read_eeprom(16)
    %Q(i2c_slave_id = #{ee[0]}
stepper_current = [#{self.class.bytes_to_number(ee[1,2])}, #{self.class.bytes_to_number(ee[3,2])}]
start_current = [#{self.class.bytes_to_number(ee[5,2])}, #{self.class.bytes_to_number(ee[7,2])}]
holding_current = [#{self.class.bytes_to_number(ee[9,2])}, #{self.class.bytes_to_number(ee[11,2])}]
stepping_mode = #{ee[13]}
controller_mode = #{ee[14]}
crc_mode = #{ee[15]}
)
  end

  def do_it(cmd)
    if @@debug
      cmd.each_byte do |byte|
        $stdout.print "#{byte} "
      end
      $stdout.print "[#{self.class.crc8(cmd)[0]}]"
    end
    write('!#' + cmd + self.class.crc8(cmd))
    sleep(0.2)
    read
  end

  def self.crc8(string)
   crc8 = 0
   magic_constant = 0x8c # can anyone explain this to me?
   string.each_byte do |byte_value|
     8.times do
        j = 1 & (byte_value ^ crc8)
        crc8 = (crc8 / 2) & 0xff
        byte_value = (byte_value / 2) & 0xff
        if j != 0
          crc8 = crc8 ^ magic_constant
        end
     end
   end
   crc8.chr
  end

  def self.number_to_bytes(number)
    (number & 0xff).chr + (number>>8&0xff).chr
  end

  def self.bytes_to_number(bytes)
    bytes[0]|bytes[1]<<8
  end

  def self.steps_per_sec(num)
    1000 / (num + 1)
  end
end

